### ISSUE TITLE
ONNX pointwise convolution and group convolution are too slow

### ISSUE THREAD

**Issue by Hyunseok-Kim0** [microsoft/onnxruntime#9192]
**Describe the bug**
I am trying to use onnx converted MobileNet-v3 on raspberry pi. However, the converted model took too much time to inference. 
To find out which part delaying the whole process, I have measured the inference time of each type of convolution operation and the result is shown below.
 
|pytorch cpu result       |conv1x1|conv3x1|conv1x3 |gconv3x3|conv3x3|
| --- | --- | --- | --- | --- | --- |
|processing time [sec]  |3.195	   |2.319	 |2.530	|2.029	 |5.715	|
|vs 3x3                          |0.559 	   |0.406	 |0.443	|0.355	 |1.000	|
|theoretical complexity|0.111	   |0.333	 |0.333	|0.016	 |1.000	|

|onnx result                  |conv1x1|conv3x1|conv1x3 |gconv3x3 |conv3x3|
| --- | --- | --- | --- | --- | --- |
|processing time [sec]  |0.869	   |1.432	 |1.613	|0.793	|3.655	|
|vs 3x3                          |0.238	   |0.392	 |0.441	|0.217	|1.000	|
|theoretical complexity|0.111	   |0.333	 |0.333	|0.016	|1.000	|

This result is gained using ubuntu desktop. as stated above, I think there is some kind of problem in the inference speed of pointwise convolution and group convolution (depthwise convoluton). The theoretical complexity of group convolution is 0.016, since the number of filter (=group) is 64 in this case.
 
Furthermore, I got similar result when I run same code on raspberry pi 4.

|pytorch cpu result|conv1x1|conv3x1|conv1x3 |gconv3x3|conv3x3|
| --- | --- | --- | --- | --- | --- |
|processing time [sec]|12.509   |17.770	 |21.508	|18.913	 |19.605	|
|vs 3x3 |0.638 	   |0.906	 |1.097	|0.965	 |1.000	|
|theoretical complexity|0.111	   |0.333	 |0.333	|0.016	 |1.000	|

|onnx result                  |conv1x1|conv3x1|conv1x3 |gconv3x3 |conv3x3|
| --- | --- | --- | --- | --- | --- |
|processing time [sec]  |2.839	   |5.292	 |5.534	|6.150	|16.651	|
|vs 3x3                          |0.171	   |0.318	 |0.332	|0.369	|1.000	|
|theoretical complexity|0.111	   |0.333	 |0.333	|0.016	|1.000	|

Pointwise convolution worked much faster in this case. However, group convolution is too slow in this case.
Onnxruntime version, pytorch version, python package versions are identical for both cases. I couldn't find any reason why group convolution is slow in most of cases, and why pointwise convolution runs faster in raspberry pi.

Please help me to solve this problem.

Thank you in advance.

**Urgency**
None.

**System information**
- Linux Ubuntu 18.04, Raspbian 10 (buster)
- ONNX Runtime installed from (source or binary): binary
- ONNX Runtime version: 1.8.1
- Pytorch version: 1.9.0
- Python version: 3.7.3

**To Reproduce**
- Describe steps/code to reproduce the behavior.

I used code attached below.

        import torch
        import torch.nn as nn
        from torch.autograd import Variable
        import time
        import matplotlib.pyplot as plt
        import onnxruntime
        import torch.backends.cudnn as cudnn

        def get_conv(filter_num, param):
            if param["sep"]:
                conv = nn.Conv2d(filter_num, filter_num, param["kernel_size"],
                                 groups=filter_num, padding=param["padding"], bias=False, dilation=param["dilation"])
            else:
                conv = nn.Conv2d(filter_num, filter_num, param["kernel_size"],
                                 groups=1, padding=param["padding"], bias=False, dilation=param["dilation"])
        
            return conv
        
        image_size = 32
        batch_num = 32
        layer_num = 32
        filter_nums = [8, 16, 32, 64]
        use_gpu = False
        
        params = {
            "conv3x3": {"kernel_size": 3, "sep": False, "padding": 1, "dilation": (1, 1)},
            "gconv3x3": {"kernel_size": 3, "sep": True, "padding": 1, "dilation": (1, 1)},
            "conv1x1": {"kernel_size": 1, "sep": False, "padding": 0, "dilation": (1, 1)},
            "conv1x3": {"kernel_size": (1, 3), "sep": False, "padding": (0, 1), "dilation": (1, 1)},
            "conv3x1": {"kernel_size": (3, 1), "sep": False, "padding": (1, 0), "dilation": (1, 1)},
        }
        
        results = {}
        results_onnx = {}
        
        for name, param in params.items():
            timings = []
            timings_onnx = []
        
            for filter_num in filter_nums:
        
                layers = [get_conv(filter_num, param) for _ in range(layer_num)]
                model = nn.Sequential(*layers)
                # model = model.to(memory_format=torch.channels_last)
                with torch.no_grad():
                    if use_gpu:
                        cudnn.benchmark = True
                        cudnn.fastest = True
                        input = Variable(torch.randn(batch_num, filter_num, image_size, image_size).cuda())
                        model.cuda()
                    else:
                        input = Variable(torch.randn(batch_num, filter_num, image_size, image_size))
        
                    # input = input.to(memory_format=torch.channels_last)
                    out = model(input)
        
                    start = time.time()
        
                    for i in range(100):
                        out = model(input)
        
                    elapsed_time = time.time() - start
                    timings.append(elapsed_time)
        
                model.eval()
                torch.onnx.export(model, (input,), "temp.onnx", opset_version=11)
                ort_session = onnxruntime.InferenceSession("temp.onnx")
                ort_inputs = {ort_session.get_inputs()[0].name: input.numpy()}
        
                ort_session.run(None, ort_inputs)
                start = time.time()
                for i in range(100):
                    ort_outs = ort_session.run(None, ort_inputs)
                elapsed_time = time.time() - start
                timings_onnx.append(elapsed_time)
        
            results[name] = timings
            results_onnx[name] = timings_onnx
        
        for name in params.keys():
            plt.plot(filter_nums, results[name], label=name, marker="o")
        
        plt.xlabel("number of filters")
        plt.ylabel("processing time")
        plt.legend()
        plt.savefig("conv_benchmark.png")
        
        plt.clf()
        
        for name in params.keys():
            plt.plot(filter_nums, results_onnx[name], label=name, marker="o")
        
        plt.xlabel("number of filters")
        plt.ylabel("processing time")
        plt.legend()
        plt.savefig("conv_benchmark_onnx.png")
        
        convs = ["conv1x1", "conv3x1", "conv1x3", "gconv3x3", "conv3x3"]
        
        print("|                      |" + "|".join(convs) + "|")
        print("|processing time [sec] |" + "|".join(["{:0.3f}\t".format(results[n][-1]) for n in convs]) + "|")
        print("|vs 3x3                |" + "|".join(
            ["{:0.3f}\t".format(results[n][-1] / results["conv3x3"][-1]) for n in convs]) + "|")
        print("|theoretical complexity|0.111\t|0.333\t|0.333\t|0.016\t|1.000\t|")
        print("|onnx result           |" + "|".join(convs) + "|")
        print("|processing time [sec] |" + "|".join(["{:0.3f}\t".format(results_onnx[n][-1]) for n in convs]) + "|")
        print("|vs 3x3                |" + "|".join(
            ["{:0.3f}\t".format(results_onnx[n][-1] / results_onnx["conv3x3"][-1]) for n in convs]) + "|")
        print("|theoretical complexity|0.111\t|0.333\t|0.333\t|0.016\t|1.000\t|")

**Comment by yufenglee** [comment]
@Hyunseok-Kim0, pointwise conv looks fine. We are working on adding a special kernel for depthwise Conv for fp32. Depthwise is expected to improve a lot with it.

**Comment by vasanthzoh** [comment]
Hi, I'm still facing slowness when doing depthwise + pointwise Conv1D vs standard Conv1D. Would love to hear an update on the depthwise kernel. Thank you very much in advance. @yufenglee

**Comment by vasanthzoh** [comment]
Could you please provide information on why this is closed as not planned? @snnn @yufenglee
